File: project/backend-services/sharespot-device-records-backend/infrastructure/persistence/postgres/src/main/java/sharespot/services/devicerecordsbackend/infrastructure/persistence/postgres/mapper/RecordMapper.java
package sharespot.services.devicerecordsbackend.infrastructure.persistence.postgres.mapper;

import sharespot.services.devicerecordsbackend.domain.model.records.*;
import sharespot.services.devicerecordsbackend.infrastructure.persistence.postgres.model.DeviceRecordEntryPostgres;
import sharespot.services.devicerecordsbackend.infrastructure.persistence.postgres.model.DeviceRecordEntryTypePostgres;
import sharespot.services.devicerecordsbackend.infrastructure.persistence.postgres.model.DeviceRecordsPostgres;

import java.util.List;
import java.util.stream.Collectors;

public class RecordMapper {

    public static DeviceRecordsPostgres domainToPostgres(DeviceRecords records) {
        var postgres = new DeviceRecordsPostgres();
        postgres.deviceId = records.getDeviceId().value();
        postgres.entries = records.getRecords().entries().stream().map(e -> {
            var entry = new DeviceRecordEntryPostgres();
            if (e instanceof BasicRecordEntry) {
                entry.type = DeviceRecordEntryTypePostgres.basic();
            } else {
                entry.type = DeviceRecordEntryTypePostgres.sensorData();
            }
            entry.content = e.getContent();
            entry.label = e.getLabel();
            return entry;
        }).collect(Collectors.toSet());
        return postgres;
    }

    public static DeviceRecords postgresToDomain(DeviceRecordsPostgres records) {
        List<RecordEntry> collect = records.entries.stream().map(e -> {
            if (e.type.equals(DeviceRecordEntryTypePostgres.basic())) {
                var label = SensorDataRecordLabel.give(e.label);
                return new SensorDataRecordEntry(label, e.content);
            } else {
                return new BasicRecordEntry(e.label, e.content);
            }
        }).collect(Collectors.toList());

        return new DeviceRecords(new DeviceId(records.deviceId), new Records(collect));
    }
}

File: project/frontend/sharespot-device-records-frontend/project/src/app/mappers/RecordMapper.ts
import {DeviceRecordDTO, RecordTypeDTO, SensorDataRecordLabelDTO} from "../dtos/RecordsDTO";
import {DeviceRecord} from "../model/DeviceRecord";
import {RecordEntry} from "../model/RecordEntry";
import {RecordType} from "../model/RecordType";
import {SensorDataRecordLabel} from "../model/SensorDataRecordLabel";

export class RecordMapper {

  static dtoToModel(dto: DeviceRecordDTO): DeviceRecord {
    let entries = dto.entries.map(e => {
      if (e.type == RecordTypeDTO.BASIC) {
        return new RecordEntry(e.label, e.content, RecordType.BASIC)
      } else {
        if ((e.label as SensorDataRecordLabelDTO) === SensorDataRecordLabelDTO.GPS_LONGITUDE) {
          return new RecordEntry(SensorDataRecordLabel.GPS_LONGITUDE, e.content, RecordType.SENSOR_DATA)
        } else {
          return new RecordEntry(SensorDataRecordLabel.GPS_LATITUDE, e.content, RecordType.SENSOR_DATA)
        }
      }
    })
    return new DeviceRecord(dto.deviceId, entries);
  }

  static modelToDto(model: DeviceRecord): DeviceRecordDTO {
    let entries = model.entries.map(e => {
      if (e.type == RecordType.BASIC) {
        return {label: e.label, content: e.content, type: RecordTypeDTO.BASIC}
      } else {
        if ((e.label as SensorDataRecordLabel) === SensorDataRecordLabel.GPS_LONGITUDE) {
          return {label: SensorDataRecordLabelDTO.GPS_LONGITUDE, content: e.content, type: RecordTypeDTO.SENSOR_DATA}
        } else {
          return {label: SensorDataRecordLabelDTO.GPS_LATITUDE, content: e.content, type: RecordTypeDTO.SENSOR_DATA}
        }
      }
    })
    return {deviceId: model.deviceId, entries: entries}
  }
}
